### PR TITLE
Configuration property loader as service to allow YAML loader

### PR DESCRIPTION
**Pull request by Git5000**
### Description

The Java property configuration loader is made more generic and as service so that user loaders (YAML, JSON) become possible. Besides the ConfigurationLoader interface and its implementation this requires public access to some classes and methods.

Linked issue: #114 

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Code style follows the tinylog standard
- [x] All classes and methods have Javadoc
- [x] Changes are covered by JUnit tests including corner cases, errors, and exception handling
- [x] Maven build works including compiling, tests, and checks (`mvn verify`)
- [x] Changes are committed by a verified email address that is assigned to the GitHub account (https://github.com/settings/emails)

### Documentation

Additions or amendments for the [public documentation](https://github.com/pmwmedia/tinylog/wiki/Documentation):

```markdown
It has API effect with the new configuration loader service.
```

### Agreements

- [x] I agree that my changes will be published under the terms of the [Apache License 2.0](https://github.com/pmwmedia/tinylog/blob/v2.0/license.txt)
- [x] I agree that my GitHub user name will be published in the release notes

### CODE REVIEW

**Comment by Git5000** [comment]
Please have a look at the code changes when you have time. I expect a few iterations on the code to become stable as it is quite a few lines of code changes although many are just copying methods from one class to another.

**Comment by Git5000** [comment]
PS: There is a design decision in the `Configuration` which can be discussed. I made the standard `PropertyConfigurationLoader` not as service but load this manually as default if no other service is found. In this way you do not need a service loader text file for all standard use cases and only one for the add ons (which will be the exception case anyway). However, to "simulate" the usage of a service loader it is necessary to make the class name and simple class name comparison by hand. I had it first also as a service loader but did not like it. It also enforced a lot of service files in the other modules like the unit tests. But have a look at the code and you will see it.

**Comment by pmwmedia** [comment]
Thank you for your pull request! The implementation looks very great and well designed.

I have just a few remarks:

1. Could you rename `PropertyConfigurationLoader` into `PropertiesConfigurationLoader`? The reason is that this loader loads **properties** files.

2. In `PropertyConfigurationLoader.load()`, the lines 104-118 can be moved into `Configuration.load()`. I think we need this functionality (merging system properties as well as resolving system properties and environment variables) also for custom configuration loaders.

3. For ProGuard, we need a fake load for `ConfigurationLoader` via `java.util.ServiceLoader` [(example)](https://github.com/pmwmedia/tinylog/blob/6abe04fe3d7e6232ccd8e71905fb368f3386ba64/tinylog-impl/src/main/java/org/tinylog/core/ConfigurationParser.java#L125)

>I had it first also as a service loader but did not like it. It also enforced a lot of service files in the other modules like the unit tests. But have a look at the code and you will see it.

What happens if you register the service file in `tinylog-api`? Won't this single service file loaded by all modules including JUnit tests?

**Comment by Git5000** [comment]
>What happens if you register the service file in tinylog-api? Won't this single service file loaded by all modules including JUnit tests

I don't know the library as well as you do. So it probably does :-) 
Do you prefer it as service then? It saves the query for the class name but adds a check for the loaders loaded as you still have to check for an instance of the standard loader e.g. As you allow one other and the standard loader to be present but not two others. I am neutral on this as I tried both and none appeared drastically better.

In 1) you are totally right of course. I will change that.

2+3) yes we can do that.

**Comment by Git5000** [comment]
The above commit has fixes for your findings 1-3 in. Please check these.

You are right, it is much better to have the general methods for the property merging and resolving separate.

For the service loading vs construction of the standard loader I wait for your decision which you prefer to have build in, i.e. it is not changed in this commit.

**Comment by Git5000** [comment]
The Proguard check removed the nice 100% test coverage. I like coverage, so I added a testcase for this.

**Comment by Git5000** [comment]
Ok, I have to look into that unittest. Somehow it is not working as expected. Do you see why it does not mock the process id? The assert in the code below works but not the "if" in the real code in `Configuration.load()`.
```

	@Test
	@PrepareForTest({Configuration.class, RuntimeProvider.class})	
	public void checkProguardHack() throws Exception {
		PowerMockito.mockStatic(RuntimeProvider.class);
		PowerMockito.mockStatic(Configuration.class);
		Mockito.when(RuntimeProvider.getProcessId()).thenReturn(Long.MIN_VALUE);
		Whitebox.invokeMethod(Configuration.class, "load");
		assertThat(RuntimeProvider.getProcessId()).isEqualTo(Long.MIN_VALUE);
	}
```

**Comment by pmwmedia** [comment]
> The Proguard check removed the nice 100% test coverage. I like coverage, so I added a testcase for this.

Wow! You are amazing! I really like and take care for code coverage. Well tested code is important for stability.

> For the service loading vs construction of the standard loader I wait for your decision which you prefer to have build in, i.e. it is not changed in this commit.

I prefer service loading as long as it doesn't require (much) more code than your current approach.

**Comment by Git5000** [comment]
I think you are right with the configuration design and I therefore changed the code back to using only the service loader. It is not more (maybe even less code) so that is ok. You tip with the location of the service file also solved the dependency issues in the tests. So I think it works fine.

There is unfortunately a complaint from the deepcode-ci-bot. But I think this is a new rule because I don't think the logging principle was changed. Could you please have a look at it.

If I understand the deepcode issue right, then I don't fully agree with it. We should not really need an 'if' guard around the internal logger. However, the issue can't stay of course. So to my understanding either the ci-bot rule needs to be changed to allow the internal logger or the logging does have to be changed. Could you have a look at this?

**Comment by pmwmedia** [comment]
> There is unfortunately a complaint from the deepcode-ci-bot. But I think this is a new rule because I don't think the logging principle was changed. Could you please have a look at it.

I have uninstalled the deepcode-ci-bot. Actually, this bot has never found anything useful. For future commits, it won't run anymore.

File: tinylog-api/src/main/java/org/tinylog/configuration/PropertiesConfigurationLoader.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+				try {
+					stream.close();
+				} catch (IOException ex) {
+					// Do nothing
+				}
+			}
+		}
+
+		Configuration.mergeSystemProperties(properties);
+		Configuration.resolveProperties(properties, EnvironmentVariableResolver.INSTANCE, SystemPropertyResolver.INSTANCE);
```
Comment (pmwmedia) [inline tinylog-api/src/main/java/org/tinylog/configuration/PropertiesConfigurationLoader.java]: I propose to call both methods in `Configuration.load()`. In this case, all future configuration loaders will benefit from merging and resolving properties automatically.

**Review by pmwmedia — CHANGES_REQUESTED** [review]

File: tinylog-api/src/main/java/org/tinylog/configuration/PropertiesConfigurationLoader.java (hunk 1 above)
Comment (Git5000) [reply tinylog-api/src/main/java/org/tinylog/configuration/PropertiesConfigurationLoader.java]: Yes you are right. This makes sense.

**Comment by Git5000** [comment]
I moved the property calls. Please have a look.

**Review by pmwmedia — APPROVED** [review]

**Comment by pmwmedia** [comment]
Thank you very much! Merged with pleasure :)
